custom_components/waste_collection_schedule/waste_collection_schedule/source/avl_ludwigsburg.py

```diff
@@ -0,0 +1,143 @@
+
+        # Creat the headers and cookies
+        headers = {
+            'Connection': 'keep-alive',
+            'Cache-Control': 'max-age=0',
+            'sec-ch-ua': '" Not A;Brand";v="99", "Chromium";v="98", "Opera";v="84"',
+            'sec-ch-ua-mobile': '?0',
+            'sec-ch-ua-platform': '"Windows"',
+            'Upgrade-Insecure-Requests': '1',
+            'Origin': 'https://www.avl-ludwigsburg.de',
+            'Content-Type': 'application/x-www-form-urlencoded',
+            'User-Agent': 'Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36 OPR/84.0.4316.21',
+            'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
+            'Sec-Fetch-Site': 'same-origin',
+            'Sec-Fetch-Mode': 'navigate',
+            'Sec-Fetch-User': '?1',
+            'Sec-Fetch-Dest': 'document',
+            'Referer': 'https://www.avl-ludwigsburg.de/privatkunden/termine/abfallkalender/suche/',
+            'Accept-Language': 'de-DE,de;q=0.9',
+        }
+        cookies = {
+            'agree': 'false',
+            'track_ga': 'false',
+        #    '_ga': 'GA1.2.783849652.1646596236',
+        #    '_gid': 'GA1.2.801427368.1646596236',
+        }
+
+        # Get the final data
+        response = s.post(URL, data=data, headers=headers, cookies=cookies)
+
+        # Stop if something else as status code 200 is returned
+        if response.status_code != 200:
+            raise Exception(
+                "Error: failed to fetch url: {}".format(
+                    URL
+                )
+            )
+
+        if response.text.find("Ort konnte nicht gefunden werden.") != -1:
+            raise Exception(
+                "Error: Ort konnte nicht gefunden werden."
+            )
+
+        if response.text.find("Straße konnte nicht gefunden werden.") != -1:
+            raise Exception(
+                "Error: Ort konnte nicht gefunden werden."
+            )
+
+        if response.text.find(".ics") == -1:
+            raise Exception(
+                "Error: No ics link found."
+            )
+
+        soup = BeautifulSoup(response.text, features="html.parser")
+        downloads = soup.find_all("a", href=True)
+        ics_link = ""
+        for download in downloads:
+            link = download.get("href")
+            if ".ics" in link:
+                ics_link = link
+        full_url = ("https://www.avl-ludwigsburg.de" + ics_link)
+        return self.fetch_ics(full_url)
+
+    def fetch_ics(self, url):
+        r = requests.get(url, headers={
+            "Referer": URL
+        })
+
+        if not r.ok:
+            raise Exception(
+                "Error: failed to fetch url: {}".format(
+                    url
+                )
+            )
+        
+        # Parse ics file, fix broken encoding
+        if r.encoding=="ISO-8859-1":
```

Comment:
Would r.encoding = "utf-8" also work? (Instead of the if else). I used this many times for other sources.

Reply:
I tried it without that, didn't work. For some reason the webservice or the python requests module received ISO-8859-1 encoded data.

Reply:
I'm not sure whether there is a misunderstanding. Therefore I repeat my point:

I know that python-requests sometimes has problems to detect the right encoding. But instead of the if/else statements, for me always a simple `r.encoding = "utf-8"` worked.

There are some examples in other sources, e.g.:

https://github.com/mampfes/hacs_waste_collection_schedule/blob/96f536ac74dc5d4acc1e3dbcb05879e47fcf23dd/custom_components/waste_collection_schedule/waste_collection_schedule/source/abfall_io.py#L127

Would that also work for you?

Reply:
This might fix it for some cases but it assumes the format of input data which are in someone elses reponsibility. Therefore I used a more flexibel approach with 570fafb8d6ee72e27db6fa6178d10df7f9914348 to be prepared for different scenarios, without assuming the encoding of the data.